File: web/src/api/earn.ts
import type { Vault, Position, VaultFilters } from '../types';
import { calculateRiskScore } from '../utils/riskScore';

const EARN_BASE_URL = import.meta.env.VITE_EARN_BASE_URL || 'https://earn.li.fi';

function enrichVault(raw: Record<string, unknown>): Vault {
  const vault: Vault = {
    id: String(raw.id ?? raw.vaultId ?? raw.address ?? ''),
    name: String(raw.name ?? raw.displayName ?? ''),
    protocol: String(raw.protocol ?? raw.protocolId ?? ''),
    chain: String(raw.chain ?? raw.chainId ?? raw.network ?? ''),
    token: String(raw.token ?? raw.tokenSymbol ?? raw.asset ?? ''),
    apy: Number(raw.apy ?? raw.apyBase ?? raw.currentApy ?? 0),
    tvl: Number(raw.tvl ?? raw.tvlUsd ?? raw.totalValueLocked ?? 0),
    tvlChange7d: raw.tvlChange7d !== undefined ? Number(raw.tvlChange7d) : undefined,
    apyHistory: Array.isArray(raw.apyHistory)
      ? (raw.apyHistory as number[])
      : undefined,
  };
  vault.riskScore = calculateRiskScore(vault);
  return vault;
}

export async function fetchVaults(filters?: VaultFilters): Promise<Vault[]> {
  const params = new URLSearchParams();
  if (filters?.chain) params.set('chain', filters.chain);
  if (filters?.token) params.set('token', filters.token);
  if (filters?.protocol) params.set('protocol', filters.protocol);

  const url = `${EARN_BASE_URL}/v1/opportunities?${params.toString()}`;

  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json' },
  });

  if (!res.ok) {
    throw new Error(`Earn API error ${res.status}: ${await res.text()}`);
  }

  const data: unknown = await res.json();

  let items: Record<string, unknown>[] = [];
  if (Array.isArray(data)) {
    items = data as Record<string, unknown>[];
  } else if (data && typeof data === 'object') {
    const obj = data as Record<string, unknown>;
    const candidate = obj.opportunities ?? obj.vaults ?? obj.data ?? obj.items;
    if (Array.isArray(candidate)) {
      items = candidate as Record<string, unknown>[];
    }
  }

  let vaults = items.map(enrichVault);

  // Client-side filters
  if (filters?.minApy !== undefined) {
    vaults = vaults.filter((v) => v.apy >= (filters.minApy as number));
  }
  if (filters?.maxRisk !== undefined) {
    vaults = vaults.filter(
      (v) => (v.riskScore ?? 0) <= (filters.maxRisk as number),
    );
  }

  return vaults;
}

export async function fetchPositions(walletAddress: string): Promise<Position[]> {
  if (!walletAddress) return [];

  const url = `${EARN_BASE_URL}/v1/positions?walletAddress=${encodeURIComponent(walletAddress)}`;

  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json' },
  });

  if (!res.ok) {
    throw new Error(`Earn API error ${res.status}: ${await res.text()}`);
  }

  const data: unknown = await res.json();

  let items: Record<string, unknown>[] = [];
  if (Array.isArray(data)) {
    items = data as Record<string, unknown>[];
  } else if (data && typeof data === 'object') {
    const obj = data as Record<string, unknown>;
    const candidate = obj.positions ?? obj.data ?? obj.items;
    if (Array.isArray(candidate)) {
      items = candidate as Record<string, unknown>[];
    }
  }

  return items.map((raw) => {
    const vaultRaw = (raw.vault ?? raw.opportunity ?? {}) as Record<string, unknown>;
    const vault = enrichVault({ ...vaultRaw, ...(raw.vaultId ? { id: raw.vaultId } : {}) });
    return {
      vaultId: vault.id,
      vault,
      walletAddress,
      balance: Number(raw.balance ?? raw.amount ?? 0),
      balanceUsd: Number(raw.balanceUsd ?? raw.valueUsd ?? raw.amountUsd ?? 0),
      shareOfPool: Number(raw.shareOfPool ?? raw.share ?? 0),
    } as Position;
  });
}

File: bot/api/earn.py
"""Earn API client."""
from __future__ import annotations

from typing import Any

import httpx

from config import EARN_BASE_URL


def _enrich(raw: dict[str, Any]) -> dict[str, Any]:
    """Normalise a vault dict from the Earn API."""
    return {
        "id": str(raw.get("id") or raw.get("vaultId") or raw.get("address") or ""),
        "name": str(raw.get("name") or raw.get("displayName") or ""),
        "protocol": str(raw.get("protocol") or raw.get("protocolId") or ""),
        "chain": str(raw.get("chain") or raw.get("chainId") or raw.get("network") or ""),
        "token": str(raw.get("token") or raw.get("tokenSymbol") or raw.get("asset") or ""),
        "apy": float(raw.get("apy") or raw.get("apyBase") or raw.get("currentApy") or 0),
        "tvl": float(raw.get("tvl") or raw.get("tvlUsd") or raw.get("totalValueLocked") or 0),
        "tvlChange7d": raw.get("tvlChange7d"),
        "apyHistory": raw.get("apyHistory"),
    }


async def fetch_vaults(
    chain: str | None = None,
    token: str | None = None,
    protocol: str | None = None,
) -> list[dict[str, Any]]:
    """Fetch vaults from the Earn API with optional filters."""
    params: dict[str, str] = {}
    if chain:
        params["chain"] = chain
    if token:
        params["token"] = token
    if protocol:
        params["protocol"] = protocol

    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(f"{EARN_BASE_URL}/v1/opportunities", params=params)
        resp.raise_for_status()
        data = resp.json()

    if isinstance(data, list):
        items = data
    elif isinstance(data, dict):
        items = (
            data.get("opportunities")
            or data.get("vaults")
            or data.get("data")
            or data.get("items")
            or []
        )
    else:
        items = []

    return [_enrich(v) for v in items]


async def fetch_positions(wallet_address: str) -> list[dict[str, Any]]:
    """Fetch user positions from the Earn API."""
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(
            f"{EARN_BASE_URL}/v1/positions",
            params={"walletAddress": wallet_address},
        )
        resp.raise_for_status()
        data = resp.json()

    if isinstance(data, list):
        items = data
    elif isinstance(data, dict):
        items = (
            data.get("positions")
            or data.get("data")
            or data.get("items")
            or []
        )
    else:
        items = []

    result = []
    for raw in items:
        vault_raw = raw.get("vault") or raw.get("opportunity") or {}
        vault = _enrich({**vault_raw, **({"id": raw["vaultId"]} if "vaultId" in raw else {})})
        result.append(
            {
                "vault": vault,
                "balance": float(raw.get("balance") or raw.get("amount") or 0),
                "balanceUsd": float(
                    raw.get("balanceUsd") or raw.get("valueUsd") or raw.get("amountUsd") or 0
                ),
                "shareOfPool": float(raw.get("shareOfPool") or raw.get("share") or 0),
            }
        )
    return result
